### PR TITLE
Assign new variables by standard names

### PR DESCRIPTION
**Pull request by juseg**
This pull request implements `ds.cf.assign(**standard_variables)`. The new method should:

- [ ] Assign variables by standard name as `**kwargs` or dictionary.
- [ ] Decide on the variable short names without user input (see below).
- [ ] Either handle or reject duplicate standard names (see below).
- [ ] Workaround xarray incompatibilities, currently #513.

Also needed:

- [ ] Add documentation, where should it go?
- [ ] Add tests, including name duplicates.

I would welcome opinions on the two points below.

## 1. Decide about the variable short name

I propose this algorithm:

- If variable has a short name, absent from the dataset, use it.

```python
ds.cf.assign(air_temperature=DataArray(name='tas')) -> 'tas'
```

- Otherwise, use the standard name as a short variable name.

```python
ds.cf.assign(air_temperature=DataArray()) -> 'air_temperature'
```

- If name is already used in dataset, warn user, and add trailing underscores.

```python
ds.cf.assign(
   air_temperature=DataArray(name=tas)).cf.assign(
   total_precipitation=DataArray(name=tas)) -> tas, tas_
```

- If name is taken by another variable in call, also add trailing underscores.

```python
ds.cf.assign(
   air_temperature=DataArray(name=tas),
   total_precipitation=DataArray(name=tas)) -> tas, tas_
```

## 2. When dataset already contains standard name

```python
ds = xr.Dataset()
ds = ds.cf.assign(air_temperature=0)
ds = ds.cf.assign(air_temperature=1)
```

I find it more difficult to decide what the method should do here.

- Override existing variable (same as `Dataset.assign`).
- Override but raise a warning.
- Assign new variable with same standard name.
- Assign new variable but raise a warning.
- Customize via keyword e.g. `existing: ignore, override, raise, warn`

Note: if we allow multiple variables with the same standard name, the resulting Dataset is technically valid, and `ds.cf` shows several variables associated with one standard name, while `ds.cf[standard_name] fails with a `KeyError`.

### CODE REVIEW

**Comment by dcherian** [comment]
I think the cf-xarray version of this only really makes sense when the assigned name is a standard name on one of the present variables, so (2) in your listing.

For (1), we should just forward on to Xarray, as usual.

>  if we allow multiple variables with the same standard name, the resulting Dataset is technically valid, and ds.cf shows several variables associated with one standard name, while ds.cf[standard_name] fails with a KeyError`

Yes, this is intentional. `ds.cf[standard_name]` will raise an error unless there is only one result, since that is the only way to return a DataArray. to get all, use `ds.cf[[standard_name]]`. Then you will get a dataset with all dataarrays with that standard name.

File: cf_xarray/accessor.py [hunk 1]
```diff
@@ -2633,6 +2633,54 @@ def grid_mapping_names(self) -> dict[str, list[str]]:
...
+        DataSet.assign
+        """
+
+        # initalize dictionary mapping short names to the new variables
+        variables = {}
+
+        # for each standard name and value pair
+        for standard_name, values in standard_variables.items():
+            # default to using existing short name or standard name
+            name = getattr(values, "name", standard_name)
```
Comment (dcherian) [inline cf_xarray/accessor.py:2666]: Something like 
https://github.com/xarray-contrib/cf-xarray/blob/52160a51c66edbec84f1412890887a7874dc2d25/cf_xarray/accessor.py#L2473

should be used here. So

```
mapped_name, = apply_mapper(_get_all, self._obj, standard_name, error=False, default=[standard_name])
```
 will do the translation to actual name in the dataset, if possible. Use this to create a new dictionary that you can then pass to `self._obj.assign`

Comment (juseg) [reply cf_xarray/accessor.py:2666]: Thank you for the feedback, I will be looking at this next week.
